pkg/drawio/subnetLayoutOverlap.go

```diff
@@ -0,0 +1,314 @@
+		y = midY + max(src.Height()/2, dst.Height()/2) + subnetHeight/2
+		x = midX
+	default:
+		// we collect a list of potential points on the second line, and choose the closest of them
+		// list of potential Xs:
+		potentialXs := []int{
+			midX + src.Width()/2 + subnetWidth/2,
+			midX + dst.Width()/2 + subnetWidth/2,
+			midX - src.Width()/2 - subnetWidth/2,
+			midX - dst.Width()/2 - subnetWidth/2,
+		}
+		// list of potential Ys:
+		potentialYs := []int{
+			midY + src.Height()/2 + subnetHeight/2,
+			midY + dst.Height()/2 + subnetHeight/2,
+			midY - src.Height()/2 - subnetHeight/2,
+			midY - dst.Height()/2 - subnetHeight/2,
+		}
+		// foreach potential X/Y, calculate its X/Y:
+		// we know that for two points on a line (y2-y1) = gradient*(x2-x1)
+		// the second line is vertical to the first, so the gradient == -dX/dY
+		// so (py - midY)*dY = (py - midY)*(-dX)
+		potentialPoints := []point{}
+		for _, px := range potentialXs {
+			py := midY + (midX-px)*dX/dY
+			potentialPoints = append(potentialPoints, point{px, py})
+		}
+		for _, py := range potentialYs {
+			px := midX + (midY-py)*dY/dX
+			potentialPoints = append(potentialPoints, point{px, py})
+		}
+		// find the closest point, which is outside of both squares:
+		score := max(src.Width(), dst.Width()) + max(src.Height(), dst.Height())
+		for _, point := range potentialPoints {
+			if !isPointInSquare(src, point) && !isPointInSquare(dst, point) {
+				newScore := abs(point.X-midX) + abs(point.Y-midY)
+				if newScore < score {
+					x, y = point.X, point.Y
+					score = newScore
+				}
+			}
+		}
+	}
+	line.addPoint(x, y)
+}
+
+// ////////////////////////////////////////////////////////
+func isPointInSquare(sq SquareTreeNodeInterface, p point) bool {
+	xMin, yMin := absoluteGeometry(sq)
+	xMax, yMax := xMin+sq.Width(), yMin+sq.Height()
+	return p.X >= xMin && p.X <= xMax && p.Y >= yMin && p.Y <= yMax
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// isLinesOverlap() checks if two lines overlap
+// both lines are are of the form Y = gradient*X + C
+// gradient is dy/dx
+// two points on the same line iff dx*(y2-y1) == dy*(x2-x1)
+
+func (lyO *subnetLayoutOverlap) isLinesOverlap(l1, l2 LineTreeNodeInterface) bool {
+	srcX1, srcY1 := lyO.tnCenter(l1.Src())
+	srcX2, srcY2 := lyO.tnCenter(l2.Src())
+	dstX1, dstY1 := lyO.tnCenter(l1.Dst())
+	dstX2, dstY2 := lyO.tnCenter(l2.Dst())
+	dx1, dy1 := dstX1-srcX1, dstY1-srcY1
+	dx2, dy2 := dstX2-srcX2, dstY2-srcY2
+	minX1, minY1 := min(srcX1, dstX1), min(srcY1, dstY1)
+	minX2, minY2 := min(srcX2, dstX2), min(srcY2, dstY2)
+	maxX1, maxY1 := max(srcX1, dstX1), max(srcY1, dstY1)
+	maxX2, maxY2 := max(srcX2, dstX2), max(srcY2, dstY2)
+	switch {
+	// is same gradient?
+	case dx1*dy2 != dx2*dy1,
+		// same gradient, is same graph?
+		dx1*(srcY2-srcY1) != dy1*(srcX2-srcX1),
+		// share domain?
+		(minX1 >= maxX2 || minX2 >= maxX1) && (minY1 >= maxY2 || minY2 >= maxY1):
```

Comment:
Shouldn't we have an `||` instead of a `&&`?
Two lines will not overlap if they have exactly the same Ys but non-overlapping Xs.

Reply:
since they are on the same graph, we can only check the X domain
I added the Y domain for cases that dx == 0

I changed the code, now it is more clear

Reply:
please notice that I add another corner case check:
	return (dx1 != 0 || dy1 != 0) && // src1 == dst1
		(dx2 != 0 || dy2 != 0) && // src2 == dst2

Reply:
I realized out that the code does not handle a case that one square has two different self loop
will fix it in another PR